core/src/connection/manager.rs

```diff
@@ -0,0 +1,479 @@
+        H::Handler: ConnectionHandler<
+            Substream = Substream<M>,
+            InEvent = I,
+            OutEvent = O,
+            Error = HE
+        > + Send + 'static,
+        <H::Handler as ConnectionHandler>::OutboundOpenInfo: Send + 'static,
+    {
+        let task_id = self.next_task_id;
+        self.next_task_id.0 += 1;
+
+        let (tx, rx) = mpsc::channel(4);
+        self.tasks.insert(task_id, TaskInfo { sender: tx, state: TaskState::Pending });
+
+        let task = Box::pin(Task::pending(task_id, self.events_tx.clone(), rx, future, handler));
+        if let Some(executor) = &mut self.executor {
+            executor.exec(task);
+        } else {
+            self.local_spawns.push(task);
+        }
+
+        ConnectionId(task_id)
+    }
+
+    /// Adds an existing connection to the manager.
+    pub fn add<M>(&mut self, conn: Connection<M, H::Handler>, info: Connected<C>) -> ConnectionId
+    where
+        H: IntoConnectionHandler<C> + Send + 'static,
+        H::Handler: ConnectionHandler<
+            Substream = Substream<M>,
+            InEvent = I,
+            OutEvent = O,
+            Error = HE
+        > + Send + 'static,
+        <H::Handler as ConnectionHandler>::OutboundOpenInfo: Send + 'static,
+        TE: error::Error + Send + 'static,
+        HE: error::Error + Send + 'static,
+        I: Send + 'static,
+        O: Send + 'static,
+        M: StreamMuxer + Send + Sync + 'static,
+        M::OutboundSubstream: Send + 'static,
+        C: Send + 'static
+    {
+        let task_id = self.next_task_id;
+        self.next_task_id.0 += 1;
+
+        let (tx, rx) = mpsc::channel(4);
+        self.tasks.insert(task_id, TaskInfo {
+            sender: tx, state: TaskState::Established(info)
+        });
+
+        let task: Pin<Box<Task<Pin<Box<future::Pending<_>>>, _, _, _, _, _, _>>> =
+            Box::pin(Task::established(task_id, self.events_tx.clone(), rx, conn));
+
+        if let Some(executor) = &mut self.executor {
+            executor.exec(task);
+        } else {
+            self.local_spawns.push(task);
+        }
+
+        ConnectionId(task_id)
+    }
+
+    /// Notifies all managed connections of an event.
+    ///
+    /// This function is "atomic", in the sense that if `Poll::Pending` is
+    /// returned then no event has been sent.
+    #[must_use]
+    pub fn poll_broadcast(&mut self, event: &I, cx: &mut Context) -> Poll<()>
+    where
+        I: Clone
+    {
+        for task in self.tasks.values_mut() {
+            if let Poll::Pending = task.sender.poll_ready(cx) { // (*)
+                return Poll::Pending;
+            }
+        }
+
+        for task in self.tasks.values_mut() {
+            let cmd = task::Command::NotifyHandler(event.clone());
+            match task.sender.start_send(cmd) {
+                Ok(()) => {},
+                Err(ref err) if err.is_full() => unreachable!("by (*)"),
+                Err(_) => {},
```

Comment:
As far as I know this is only reached on disconnection, right? Would this also be unreachable, given that we control both ends?

Reply:
To my understanding the receiver corresponding to `task.sender` is owned by the background task, which may in principle drop the receiver any time. How do you come to the conclusion that "we control both ends" (assuming by "we" you mean the current thread executing this function)?

Reply:
@mxinden Do you have further comments on the matter?

Reply:
@romanb sorry, I confused this with a different `start_send`. Marking as resolved.